Torch detector + scene detection without plane fitting

alex-mitrevski opened the pull request:
This PR:
* Adds a `TorchImageDetector` class to `image_detector.py`, which instantiates and runs a PyTorch object detector and classifier
* Introduces an `ObjectDetectionActionServer` class in `scene_detection_action.py`, which can be used for detecting objects without plane fitting (i.e. the result is simply a list of objects). An `object_detection_action_server` script for running the server has also been added
* Renames the `ImageDetectionActionServer` class to `PlaneDetectionActionServer`
* Renames the `image_detection_action_server` script to `plane_detection_action_server`

The documentation has been updated to reflect the changes.

Depends on PR https://github.com/b-it-bots/mas_perception_msgs/pull/7

alex-mitrevski committed:
Added a torch detector interface and an action server for detecting objects without fitting a plane

minhnh commented on ros/src/mas_perception_libs/scene_detection_action.py:
```diff
@@ -58,7 +60,142 @@ def is_object_on_plane(plane_msg, obj_box_msg, z_tolerance=0.05):
+
+    def _execute_cb(self, _):
+        # subscribe and wait for cloud message
+        print('Object detection action request received')
```
`loginfo` here for consitency?

minhnh reviewed (approved):
Looks fine to me.

minhnh commented on ros/src/mas_perception_libs/scene_detection_action.py:
```diff
@@ -58,7 +60,142 @@ def is_object_on_plane(plane_msg, obj_box_msg, z_tolerance=0.05):
+            max_coord = np.nanmax(obj_coords, axis=0)
+            mean_coord = np.nanmean(obj_coords, axis=0)
+
+            # fill object geometry info
```
could move this to a different function to reduce repetitive code

alex-mitrevski replied:
Added a function in `utils` for this.

minhnh commented on ros/src/mas_perception_libs/image_detector.py:
```diff
@@ -240,3 +242,74 @@ def process_image_msg(self, img_msg):
+
+                num_detected_objects = len(pred_boxes)
+                for i in range(len(num_detected_objects)):
+                    obj_data_dict = {ImageDetectionKey.CLASS: pred_class[i],
```
could move this to a different function to reduce repetitive code

alex-mitrevski replied:
I don't think we'd gain much by adding a function for this since we practically have a one-liner here.

minhnh replied:
yeah actually I also think this may differ for different detection libraries/algorithms (i.e. `pred_boxes` may change)

alex-mitrevski committed:
print -> rospy.loginfo for consistency

alex-mitrevski committed:
[scene_detection_action/ObjectDetection] Fixed object header

alex-mitrevski committed:
[utils] Added a fn for converting bounding box detections to object msgs

The action servers in scene_detection_action.py have been adapted to use this function.

alex-mitrevski committed:
[image_detector/TorchImageDetector] Type and error fixes

alex-mitrevski committed:
[scene_detection_action] The object detection server was setting an incorrect result

alex-mitrevski committed:
mv image_detection.launch -> plane_detection.launch

The documentation has also been updated accordingly

alex-mitrevski committed:
mv object_detector.py plane_detector.py
